File: Tutoriels/CSharp/Patrons_Conception/Creation/Singleton_Securise/src/GSingleton.cs
//===============================================
using System;
//===============================================
public class GSingleton {
    //===============================================
    private GSingleton() {
        Data = "_NO_DATA_";
    }
    //===============================================
    public static GSingleton Instance() {
        lock(m_instanceLock) {
            if(m_instance == null) {
                m_instance = new GSingleton();
            }
            return m_instance;
        }
    }
    //===============================================
    public void Print() {
        Console.WriteLine(Data);
    }
    //===============================================
    public string Data {get; set;}
    //===============================================
    private static GSingleton m_instance = null;
    private static readonly object m_instanceLock = new object();
    //===============================================
}
//===============================================

File: Tutoriels/Cpp/Patrons_Conception/code/src/manager/GSingleton.cpp
//================================================
#include "GSingleton.h"
//================================================
GSingleton* GSingleton::m_instance = 0;
//================================================
GSingleton::GSingleton() {

}
//================================================
GSingleton::~GSingleton() {
	
}
//================================================
GSingleton* GSingleton::Instance() {
	if(m_instance == 0) {
        m_instance = new GSingleton;
	}
	return m_instance;
}
//================================================
void GSingleton::setData(const string& data) {
    m_data = data;
}
//================================================
void GSingleton::print() const {
    cout << "Je suis un singleton: " << m_data << "\n";
}
//================================================
